api/v3/auth.go

```diff
@@ -0,0 +1,612 @@
+package v3
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/bobheadxi/res"
+	grpc_middleware "github.com/grpc-ecosystem/go-grpc-middleware"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"gopkg.in/dgrijalva/jwt-go.v3"
+
+	"github.com/RTradeLtd/Temporal/api/v3/proto/auth"
+	"github.com/RTradeLtd/Temporal/eh"
+	"github.com/RTradeLtd/Temporal/queue"
+	"github.com/RTradeLtd/database/models"
+)
+
+const (
+	claimUser      = "id"
+	claimChallenge = "challenge"
+	claimOrigAt    = "orig_iat"
+	claimExpiry    = "exp"
+)
+
+// JWTConfig denotes JWT signing configuration
+type JWTConfig struct {
+	Key   string
+	Realm string
+
+	Timeout     time.Duration
+	SigningAlgo jwt.SigningMethod
+}
+
+// AuthService implements TemporalAuthService
+type AuthService struct {
+	users   userManager
+	usage   usageManager
+	credits creditsManager
+	emails  publisher
+
+	jwt JWTConfig
+	dev bool
+
+	l *zap.SugaredLogger
+}
+
+// NewAuthService returns a new instance of the v3 authentication service
+func NewAuthService(
+	users userManager,
+	usage usageManager,
+	credits creditsManager,
+	emails publisher,
+
+	jwt JWTConfig,
+	dev bool,
+
+	l *zap.SugaredLogger,
+) *AuthService {
+	return &AuthService{users, usage, credits, emails, jwt, dev, l}
+}
+
+// Register returns the Temporal API status
+func (a *AuthService) Register(ctx context.Context, req *auth.RegisterReq) (*auth.User, error) {
+	var (
+		email = req.GetEmailAddress()
+		user  = req.GetCredentials().GetUsername()
+		pw    = req.GetCredentials().GetPassword()
+		l     = a.l.With("user", user, "email", email)
+	)
+
+	// validate email
+	if strings.ContainsRune(email, '+') {
```

Comment:
This reminds me as well, we should probably also validate the email is valid as well. The inability to use a `+` was due to email catch-all routing, but I've noticed people signing up with emails like `username.............` so if we get repeated sequences of a `.` its probably safe to deem the email as invalid as someone either typo'd or is just being a troll

Basically the validation rules are:
* No usage of catch-all routing (+)
* No trailing periods (ie, `username.@example.org`, `username..........@example.org`)

Reply:
Did the above, and also implement a basic domain verification - see 5d52f34

Reply:
Excellent :ok_hand: